packages/transport-webrtc/test/certificate.spec.ts

```diff
@@ -0,0 +1,94 @@
+import type { TransportCertificate } from '../src/index.js'
+import type { TransportManager } from '@libp2p/interface-internal'
+
+const crypto = new Crypto()
+
+describe('WebRTCDirectListener', () => {
+  let listener: WebRTCDirectListener
+  let components: WebRTCDirectTransportComponents
+  let init: WebRTCDirectListenerInit
+
+  beforeEach(() => {
+    components = {
+      peerId: { toB58String: () => 'QmPeerId' } as any,
+      privateKey: {} as any,
+      logger: defaultLogger(),
+      transportManager: stubInterface<TransportManager>()
+    }
+
+    init = {
+      certificateDuration: 10,
+      upgrader: {} as any
+    }
+
+    listener = new WebRTCDirectListener(components, init)
+  })
+
+  afterEach(() => {
+    sinon.restore()
+  })
+
+  it('should generate a certificate with the configured duration', async () => {
+    const keyPair = await crypto.subtle.generateKey({
+      name: 'ECDSA',
+      namedCurve: 'P-256'
+    }, true, ['sign', 'verify'])
+
+    const certificate: TransportCertificate = await generateTransportCertificate(keyPair, {
+      days: init.certificateDuration!
+    })
+
+    expect(new Date(certificate.notAfter).getTime()).to.be.closeTo(
+      new Date().getTime() + init.certificateDuration! * 86400000,
+      1000
+    )
+  })
+
+  it('should re-emit listening event when a new certificate is generated', async () => {
+    const emitSpy = sinon.spy(listener as any, 'safeDispatchEvent')
+    const generateCertificateSpy = sinon.spy(generateTransportCertificate)
+
+    await (listener as any).startUDPMuxServer('127.0.0.1', 0)
+
+    expect(generateCertificateSpy.called).to.be.true
+    expect(emitSpy.calledWith('listening')).to.be.true
+  })
+
+  it('should generate a new certificate before expiry with default threshold', async () => {
+    init.certificateExpiryThreshold = undefined
+    listener = new WebRTCDirectListener(components, init)
+    ;(listener as any).certificate = {
+      notAfter: new Date(Date.now() + 5 * 86400000).getTime()
+    }
+
+    const isCertificateExpiringSpy = sinon.spy(listener as any, 'isCertificateExpiring')
+    const generateCertificateSpy = sinon.spy(listener as any, 'createAndSetCertificate')
+
+    await (listener as any).startUDPMuxServer('127.0.0.1', 0)
+
+    expect(isCertificateExpiringSpy.returned(true)).to.be.true
+    expect(generateCertificateSpy.called).to.be.true
+  })
+
+  it('should generate a new certificate before expiry with custom threshold', async () => {
+    (listener as any).certificate = {
+      notAfter: new Date(Date.now() + 4 * 86400000).getTime()
+    }
+
+    const isCertificateExpiringSpy = sinon.spy(listener as any, 'isCertificateExpiring')
+    const generateCertificateSpy = sinon.spy(listener as any, 'createAndSetCertificate')
+
+    await (listener as any).startUDPMuxServer('127.0.0.1', 0)
+
+    expect(isCertificateExpiringSpy.returned(true)).to.be.true
+    expect(generateCertificateSpy.called).to.be.true
+  })
```

Comment:
For these tests please do not depend on internal state (e.g. private fields and methods), it makes the tests fragile and which then makes them hard to maintain.

Instead, make the `WebRTCDirectListener.certificate` field public, configure short timeouts for certificate expiry/renewal thresholds (e.g. under 100ms), use [delay](https://www.npmjs.com/package/delay) and assert on the certificate status after time has passed.